pkg/virt-controller/watch/vmi.go

```diff
@@ -254,8 +257,21 @@ func (c *VMIController) execute(key string) error {
 	if !controller.ObservedLatestApiVersionAnnotation(vmi) {
 		vmi := vmi.DeepCopy()
 		controller.SetLatestApiVersionAnnotation(vmi)
+		key := controller.VirtualMachineInstanceKey(vmi)
+		c.vmiExpectations.SetExpectations(key, 1, 0)
 		_, err = c.clientset.VirtualMachineInstance(vmi.ObjectMeta.Namespace).Update(vmi)
-		return err
+		if err != nil {
+			c.vmiExpectations.LowerExpectations(key, 1, 0)
+			return err
+		}
+		return nil
+	}
+
+	// If needsSync is true (expectations fulfilled) we can make save assumptions if virt-handler or virt-controller owns the pod
+	needsSync := c.podExpectations.SatisfiedExpectations(key) && c.vmiExpectations.SatisfiedExpectations(key)
```

Comment:
Do we have to worry about other informers like the dataVolumeInformer or pvcinformer?

Reply:
DataVolume might be important as well. I think we should have the same logic for DataVolume

Reply:
The main purpose right now is to not update the VMI when we don't see it's latest state. If other informers like pod informer, dv informer, ... fire we will not update the VMI status when we know that we don't have the latest state in cache. I think the current implementation of the PR is exactly what we want right now.

Reply:
So is the caveat here that using VMs with datavolumes will still cause collisions?  DVs might not create a lot of events, but should there be an issue created in case that ever changes?

Reply:
I think the main point is that we may need after analysis also expectations on the VM controller, but especially for the DV creation in the VM controller we already have an expectation. Further the DV informer in the VMI controller is only used to inform the controller if it can proceed with it's work. It is not updating it. The now added expectation logic automatically protects against 409s induced by the DV informer like it does for pod informers casing 409s on the VMI.

Reply:
The VM controller may hit 409s as well for similar issues that the VMI controllers hit. Really this is possible for pretty much any resource KubeVirt controls and updates where the resource's keys get queued based on watching other resources.

To make progress, I think we should tackle this on a one by one basis per resource. I think doing an analysis of the VM controller is a logical next first step.

Reply:
That works for me.  My suggestion was that we track these changes so they eventually get added to the other informers (if they apply).